File: code/python/data_structures_article-02.py
"""Binary Search Tree Heart Method (Basic Operations)"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().deleteNode(*args, **kwargs)


class _Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if not root:
            return None
        if root.val == key:
            if not root.left:
                return root.right
            if not root.right:
                return root.left
            min_node = self.getMin(root.right)
            root.right = self.deleteNode(root.right, min_node.val)
            min_node.left = root.left
            min_node.right = root.right
            root = min_node
        elif root.val > key:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        return root
    def getMin(self, node: TreeNode) -> TreeNode:
        while node.left:
            node = node.left
        return node

File: code/java/data_structures_article-02.java
public class Data_structures_article_02 {
    /**
     * Binary Search Tree Heart Method (Basic Operations)
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 2) {
            throw new IllegalArgumentException("Expected: (TreeNode root, int key)");
        }

        Solution solution = new Solution();
        return solution.deleteNode((TreeNode) args[0], (int) args[1]);
    }

    private static class Solution {

        public TreeNode deleteNode(TreeNode root, int key) {
            if (root == null) return null;
            if (root.val == key) {
                if (root.left == null) return root.right;
                if (root.right == null) return root.left;
                TreeNode minNode = getMin(root.right);
                root.right = deleteNode(root.right, minNode.val);
                minNode.left = root.left;
                minNode.right = root.right;
                root = minNode;
            } else if (root.val > key) {
                root.left = deleteNode(root.left, key);
            } else if (root.val < key) {
                root.right = deleteNode(root.right, key);
            }
            return root;
        }
        TreeNode getMin(TreeNode node) {
            while (node.left != null) node = node.left;
            return node;
        }
    }
}
